### DISCUSSION TITLE
Why would I prefer Rainbow Gum over the built-in Java 9 System.Logger?

### DISCUSSION THREAD

**Discussion by hrstoyanov** [jstachio/rainbowgum#158] [General]
Please covience me.

**Comment by agentgt** [comment]
`System.Logger` is a facade just like SLF4J and not an implementation. Rainbow Gum provides no logging API of its own on purpose. It provides backing implementations of both SLF4J and `System.Logger`. So it is more a question of why use `SLF4J` over `System.Logger`. 

The default implementation of `System.Logger` will use `java.util.logging` which is incredibly inefficient and has very few output styles (I think only file and console). Furthermore its parameter replacement uses `MessageFormat` which again is incredibly inefficient. 

That being said if one uses `System.Logger` without parameter interpolation (uses the condition checking and then concat strings) and a faster backing implementation or bridge it is mostly fine.

What happens in frameworks like Spring Boot, Micronaut or Quarkus is that they use the **System.Logger -> SLF4J bridge System Logger Bridge -> Logback/Log4j2**. You can use that approach in Rainbow Gum as well but there are caveats that are discussed here: https://jstach.io/doc/rainbowgum/current/apidocs/#system_logger

**Reply by hrstoyanov** [reply]
I read the details and I find your strong opinions matching mine! So, I will give it a try soon though the Java 9 system logger.

**Reply by agentgt** [reply]
There is a release coming that will make the System Logger implementat much faster and will check if slf4j or similar is not present and not queue.

https://jstach.io/rainbowgum/#system_logger

> The exception of the above is if [no other facade is detected](https://jstach.io/rainbowgum/io.jstach.rainbowgum/io/jstach/rainbowgum/spi/RainbowGumServiceProvider.RainbowGumEagerLoad.html) the module will initialize Rainbow Gum. The idea is that your application only uses the JDK logging facilities (the Rainbow Gum SLF4J implementation is not found in the module/classpath). Thus only having the dependencies rainbowgum-core, and rainbowgum-jdk will work as normal.

Also while it Rainbow Gum is opinionated on being modern and small it is becoming less opinionated on configuration. The readme needs to be adjusted for that.

**Reply by hrstoyanov** [reply]
I would like to avoid using slf4j , would prefer if it is not auto-detected and I have explicit control (it is inflicted on my code by other libs)

**Reply by hrstoyanov** [reply]
Also, I currently customize System logger with jvm properties for launching my Helidon 4 server - this should be picked up extremely  early in the jvm bootstrap process and is very convenient and standard way to configure JUL with no code. **Can Rainbow Gum do that**?
```
-Djava.util.logging.ConsoleHandler.formatter=java.util.logging.SimpleFormatter
-Djava.util.logging.SimpleFormatter.format="[%1$tT] [%4$-4s] %5$s%6$s%n"
```

**Reply by agentgt** [reply]
Yes the issue is that it won't show output 0.7.1 unless you tell it to eagerly load. This is fixed in `main`.

How you can in 0.7.1 tell it to eagerly load is: `-Dlogging.systemlogger.intialize=true`

["logging.systemlogger.intialize"](https://jstach.io/doc/rainbowgum/current/apidocs/io.jstach.rainbowgum.systemlogger/io/jstach/rainbowgum/systemlogger/RainbowGumSystemLoggerFinder.html#INTIALIZE_RAINBOW_GUM_PROPERTY)

In the next release this will happen automatically if something like SLF4J is not detected.

**Reply by agentgt** [reply]
So basically and I will add this to the official docs. If you only want System.Logger:

You add these deps:

* io.jstach.rainbowgum:rainbowgum-core
* io.jstach.rainbowgum:rainbowgum-jdk
* io.jstach.rainbowgum:rainbowgum-pattern (if you want pattern formatting)

Then if you are only doing console you do

```
-Dlogging.systemlogger.initialize=true
-Dlogging.appender.console.encoder=pattern
-Dlogging.encoder.console.pattern=%d{HH:mm:ss}] [%-4level] %msg%ex%n
```

That will roughly match the JUL settings you have.

**Reply by hrstoyanov** [reply]
Thank you @agentgt , will try it soon!
Just to re-iterate - I don't want SL4J used automatically/auto-detected, just because it happens to be on the classpath/modulepath. This should be under the strict control of the developer, and permitted (or disabled),  perhaps with a system property.

**Reply by agentgt** [reply]
Yes so long as you do not put

`io.jstach.rainbowgum:rainbowgum-slf4j `

in the class/modulepath you are good to go it. Do not include as a dependency that is why I listed the ones you need.

Note that

`io.jstach.rainbowgum:rainbowgum`

includes `rainbowgum-slf4j` so you don't want to do that.

(the above is Maven GAV Coordinates minus the version).


Just be aware that if any of your dependencies use SLF4J they will try to load it but it will not load Rainbow Gum however it you will get a nasty gram from SLF4J that it could not find anything.

The auto-detection I was talking about is not using reflection (well not the bad kind) but the `java.util.ServiceLoader` and even that is done as you said through a property:

https://jstach.io/rainbowgum/io.jstach.rainbowgum.systemlogger/io/jstach/rainbowgum/systemlogger/RainbowGumSystemLoggerFinder.InitOption.html 

That InitOption choices are what you feed to:

```
-Dlogging.systemlogger.initialize=
```


This isn't so much about loading SLF4J it is about whether not the `System.Logger` loads Rainbow Gum or queues and waits for something else to load.

Copy from javadoc:

[CHECK](https://jstach.io/rainbowgum/io.jstach.rainbowgum.systemlogger/io/jstach/rainbowgum/systemlogger/RainbowGumSystemLoggerFinder.InitOption.html#CHECK)
(default) Will check if there are implementations of [RainbowGumServiceProvider.RainbowGumEagerLoad](https://jstach.io/rainbowgum/io.jstach.rainbowgum/io/jstach/rainbowgum/spi/RainbowGumServiceProvider.RainbowGumEagerLoad.html) and if there are not will load rainbow gum.
[FALSE](https://jstach.io/rainbowgum/io.jstach.rainbowgum.systemlogger/io/jstach/rainbowgum/systemlogger/RainbowGumSystemLoggerFinder.InitOption.html#FALSE)
Will not initialize rainbow gum.
[REUSE](https://jstach.io/rainbowgum/io.jstach.rainbowgum.systemlogger/io/jstach/rainbowgum/systemlogger/RainbowGumSystemLoggerFinder.InitOption.html#REUSE)
Will reuse an existing rainbow gum or fail.
[TRUE](https://jstach.io/rainbowgum/io.jstach.rainbowgum.systemlogger/io/jstach/rainbowgum/systemlogger/RainbowGumSystemLoggerFinder.InitOption.html#TRUE)
Will initialize rainbow gum.


I realize it is a little confusing and I will try to improve the doc some more.

**Reply by hrstoyanov** [reply]
Thanks @agentgt - dropping the jar as means to control SL4J usage is fine. I admit I did not read the complete docs yet, but hopefully my question help with what need to be answered for newcomers first

**Reply by agentgt** [reply]
Your input is super valuable. It reminds me where there are holes in the doc that I need to fix.

Part of why Rainbow Gum is so modular is exactly for your use case but sadly I had to make the easy path for those that use SLF4J as most do.

Neither Logback, Log4J2, or even Tiny Log would let you do what we just talked about. Logback is especially bad as its core depends on SLF4J!
